### PR TITLE
kube: Return errors when users aren't allowed to list resources

### PR DESCRIPTION
**Pull request by tigrato**
Previously to this commit, if the user was denied to list resources from a certain namespace, Teleport would return an empty list instead of returning an error if it's impossible for the request to match any resource in that namespace.

This commit returns an error for cases where the role's `kubernetes_resources` denies any resource to be returned. This is a small nit to incorporate some customer expectations.

Changelog: Improve Kubernetes Access feedback when it's impossible for a `list` operation to ever return any resources based on `kubernetes_resources` defined.

### CODE REVIEW

File: lib/utils/replace.go [hunk 1]
```diff
@@ -221,6 +221,110 @@ func KubeResourceMatchesRegex(input types.KubernetesResource, resources []types.
...
+			if input.Kind != resource.Kind && resource.Kind != types.Wildcard {
+				continue
+			}
+			// if the resource is cluster-wide, the command is deny and it's a wildcard resource
+			// match all resources.
+			if isClusterWideResource && isDeny && resource.Name == types.Wildcard {
+				return true, nil
+			} else if isClusterWideResource {
+				return !isDeny, nil
+			}
```
Comment (hugoShaka) [inline lib/utils/replace.go:305]: Why do we make a special case for the wildcard name?

e.g. I run `kubectl get clusterrole/foobar`, won't I have different results for

```
deny:
  kube_resource:
    - kind: clusterrole
      name: "*"
```

and

```
deny:
  kube_resource:
    - kind: clusterrole
      name: "foobar"
```

File: lib/utils/replace.go [hunk 2]
```diff
@@ -221,6 +221,110 @@ func KubeResourceMatchesRegex(input types.KubernetesResource, resources []types.
...
+		case targetsReadOnlyNamespace && resource.Kind != types.KindKubeNamespace && resource.Namespace != "":
+			// If the user requests a read-only namespace get/list/watch, they should
+			// be able to see the list of namespaces they have resources defined in.
+			// This means that if the user has access to pods in the "foo" namespace,
+			// they should be able to see the "foo" namespace in the list of namespaces
+			// but only if the request is read-only.
+			cond := !isDeny || resource.Name == types.Wildcard && resource.Namespace == types.Wildcard
+			if input.Namespace == "" && cond {
+				return cond, nil
+			}
```
Comment (hugoShaka) [inline lib/utils/replace.go:291]: Here, input kind is a namespace (because of `targetsReadOnlyNamespace`). Is there a case in which `input.Namespace` is not `""`? I thought kube did not support nested namespaces.

File: lib/utils/replace.go (hunk 1 above)
Comment (hugoShaka) [reply lib/utils/replace.go:305]: Addendum: I think I get it now. I read the `shouldBeAllowed` part and we try to predict if the user is expected to have access to this kind/verb/namespace combo so we can raise a nice error message instead.

If this is correct, the explanation should be reflected both in the function godoc and in the function name.

Comment (tigrato) [reply lib/utils/replace.go:305]: Indeed, your observation is accurate. The objective of this function is to assess whether the user is permitted to execute the `kubectl get pods -n kube-system` command, based on the defined `kubernetes_resource` rules. The aim is to present a more user-friendly error message when there is absolute certainty that the user is not authorized to perform that operation.

Consider a user with the following `kubernetes_resources` configuration:

```yaml
allow:
  kubernetes_resources:
    - kind: pods
      name: "*"
      namespace: default
    - kind: pods
      name: "*"
      namespace: dev
deny:
    - kind: pods
      name: "really-important-service"
      namespace: default
    - kind: pods
      name: "*"
      namespace: kube-system
```

This user is granted permission to list pods in the `default` and `dev` namespaces. However, access to the `really-important-service` pod is restricted due to the deny rule in the `default` namespace. In the response for the `dev` namespace, the output mirrors that of Kubernetes. Yet, for the `default` namespace, any mention of `really-important-service` needs to be excluded.

Conversely, the user is prohibited from listing pods in the `kube-system` namespace, and Teleport should generate an error message stating: "User xyz cannot list ...". The same error should be returned for any other namespace on the list where the listing request is denied.

In summary, Teleport should only produce errors when it is certain that the deny rule will encompass all pods within the specified namespace. The function exhibits distinct behavior for allow and deny clauses. For an allow clause, a true result is returned upon any match, while for a deny clause, true is only returned if it matches every possible pod existing in the namespace.

File: lib/utils/replace.go (hunk 2 above)
Comment (tigrato) [reply lib/utils/replace.go:291]: You are correct. I copied from other location and forgot to think :disappointed:

File: lib/utils/replace.go (hunk 1 above)
Comment (hugoShaka) [reply lib/utils/replace.go:305]: Thanks for the explanation, I would rename the function to something that suggests its usage based on your explanation. For example

```golang
// KubeResourceCouldMatchRules assess whether the user is permitted to perform its request
// based on the defined kubernetes_resource rules. The aim is to catch cases when the user
// has no access and present then a more user-friendlyerror message instead of returning
// an empty list.
// This function is not responsible for enforcing access rules.
func KubeResourceCouldMatchRules() 
```

File: lib/utils/replace.go [hunk 3]
```diff
@@ -221,6 +221,109 @@ func KubeResourceMatchesRegex(input types.KubernetesResource, resources []types.
...
+		// When the resource has a wildcard verb, we only allow one verb in the
+		// resource input.
+		if !isVerbAllowed(resource.Verbs, verb) {
+			continue
+		}
+		switch {
+		// If the user has access to a specific namespace, they should be able to
+		// access all resources in that namespace.
+		case resource.Kind == types.KindKubeNamespace:
+			cond := !isDeny || resource.Name == types.Wildcard
```
Comment (AntonAM) [inline lib/utils/replace.go:267]: We already have `cond` as a parameter of this function, maybe some other name just to make sure it's not confused with the first one?

File: lib/utils/replace_test.go [hunk 4]
```diff
@@ -558,3 +558,613 @@ func TestKubeResourceMatchesRegex(t *testing.T) {
...
+					Name:  "default",
+					Verbs: []string{types.KubeVerbGet},
+				},
+			},
+			assert:  require.NoError,
+			matches: true,
+			action:  types.Allow,
+		},
+		{
+			name: "namespace denying update access to pod",
```
Comment (AntonAM) [inline lib/utils/replace_test.go:1073]: There are several cases with this same wrong name.

**Review by AntonAM — APPROVED** [review]
Nice!

**Review by hugoShaka — APPROVED** [review]

**Comment by tigrato** [comment]
@zmb3 @r0mant 

can I have a `/excludeflake *`?
